packages/client/hooks/useSortSubColumns.ts

```diff
@@ -0,0 +1,48 @@
+import {useLayoutEffect, useEffect} from 'react'
+import {commitLocalUpdate} from 'react-relay'
+import {SubColumn} from '~/types/constEnums'
+import {GroupingKanbanColumn_reflectionGroups} from '~/__generated__/GroupingKanbanColumn_reflectionGroups.graphql'
+import useAtmosphere from './useAtmosphere'
+
+const useSortSubColumns = (
+  isWidthExpanded: boolean,
+  reflectionGroups: GroupingKanbanColumn_reflectionGroups
+) => {
+  const atmosphere = useAtmosphere()
+  useLayoutEffect(() => {
+    if (!reflectionGroups) return
+    commitLocalUpdate(atmosphere, (store) => {
+      reflectionGroups.forEach((group, index) => {
+        const reflectionGroup = store.get(group.id)
+        if (!reflectionGroup) return
+        const subColumn = index % 2 === 0 ? SubColumn.LEFT : SubColumn.RIGHT
+        reflectionGroup.setValue(subColumn, 'subColumn')
+      })
+    })
+  }, [isWidthExpanded])
+
+  const updateReflectionGroups = () => {
+    if (!isWidthExpanded || !reflectionGroups) return
+    commitLocalUpdate(atmosphere, (store) => {
+      let leftSubColumnCount = 0
+      let rightSubColumnCount = 0
+      reflectionGroups.forEach((group) => {
+        const reflectionGroup = store.get(group.id)
+        if (!reflectionGroup) return
+        const currentSubColumn = reflectionGroup.getValue('subColumn')
+        if (!currentSubColumn) {
+          const subColumn =
+            leftSubColumnCount > rightSubColumnCount ? SubColumn.RIGHT : SubColumn.LEFT
+          reflectionGroup.setValue(subColumn, 'subColumn')
+        } else {
+          currentSubColumn === SubColumn.LEFT
+            ? (leftSubColumnCount += 1)
+            : (rightSubColumnCount += 1)
+        }
+      })
+    })
+  }
+  useEffect(() => updateReflectionGroups(), [reflectionGroups])
```

Comment:
this gets a little spicy because the dependency is `reflectionGroups` and the function actually changes `reflectionGroups` which means we're guaranteed to trigger at least 2 re-renders.

big picture, the question is, "is the subColumn state or a calculation?" If it's a calculation, then we can accomplish this without requiring the relay store & doing any re-rendering:

- determine a `columnCount` like how many columns we can support (1, 2, 3, etc.)
- for each column, do something like `reflectionGroups.map((group, index) =>  if (index %  columnCount !== 0 return null...`

Reply:
Initially, I used a calculation instead of state but this caused reflection groups that the user isn't updating to move around. For example, in the recording below I move group 3 into group 2. This removes group 3 which updates the indexes of subsequent groups meaning `idx % columnCount` logic causes group 5 to move. 

So I think we have to use state to keep track of the original position of the group. To support more than two columns, I could keep track of the subColumn index rather than `LEFT` `RIGHT`?


https://user-images.githubusercontent.com/39854876/103781633-fda1c880-500c-11eb-918c-8936d7de67ba.mp4

Reply:
valid point! 
what if we moved the handling outside of the `useEffect` and put it next to where `setIsWidthExpanded` is called?

So big picture:
- you click the expand button
- a calculation is performed to see how many extra columns we can add (minimum 2 cols, because we want expand to do _something_)
- a `subColumnCount` state is set with an integer (1,2,3, etc) 
- a `subColumn` integer (0, 1, 2) value is assigned to each child in that column

doing it on click should allow us to set state outside of the `useEffect` hook

Reply:
I've pushed an update where it calculates the subColumn index when clicking the expand button so we can now have 3+ subColumns 🎉 
There is a still a `useEffect` listening to `reflectionGroup` changes. I think we still need this as it sorts a new reflection group that doesn't have a subColumn index. This happens when a user adds a new reflection or removes a reflection from a group.